Async report feature

tomdickman opened the pull request:
This feature offloads the responsibility of running the flash apocalypse report to a cron job, storing the latest results in a new table, and loading paginated results from the table using a renderer when an admin goes to the report.
A lot of the logic has been shifted into classes handling the model (audit_activity) or controlling the data which is sent to the renderer (audit_manager.)

tomdickman commented:
Updated to fix issues identified in CI tests and removed the now superfluous locallib.

danmarsden commented:
Thanks for working on this - will be nice to have this performing for some of the larger sites!

In general the code looks pretty clean but here are a few general comments - I will post some other comments inline on the PR.

1) It would be better to store the integers courseid/activityid etc in the report_apocalypse table - pretty much the same data that the original SQL returns - that way the table can also be used in other reports etc. and then mapping these to urls/names/etc on display of the report. One of the issues with the way you currently do this is that on every insert to the table, you make a db call to get the course category information.

2) I'm not a fan of unnecessary abstraction... as Plugin Guardian on moodle.org I review a lot of 3rd party code, and it can be really easy to get carried away and create functions/classes that aren't really needed when the code should really just sit in-line. It can be really easy to turn 1 or 2 lines of easy to read code, and then put them into 8 lines (or more) hard to read code - here are some examples:
- the new apocalypse_datetime class is overkill - putting this into a class doesn't really give any tangible benefit.
- get_all_modules does not need to be a function, it's only used once, just call the Moodle API query inline where it's needed instead of wrapping it in a function call.
- get_course_url_from_record/get_category_from_record probably don't need to be in functions - they are only called once.
- does audit_manager really need to have a parent class that does nothing? - can't it just be a simple single class?

danmarsden commented on classes/audit_manager.php:
```diff
@@ -0,0 +1,376 @@
+     * @return \moodle_recordset A moodle_recordset instance
+     * @throws \dml_exception
+     */
+    public static function get_audit_results() {
```
this shouldn't be a function - you also need to make sure that when you use recordsets you must close them in the same function - you can't open a recordset in one function and then close it in another... but this really shouldn't be a function on it's own in the first place.

tomdickman replied:
Removed.

danmarsden commented on classes/audit_manager.php:
```diff
@@ -0,0 +1,376 @@
+        global $DB;
+
+        $activity = self::build_audit_activity_from_record($record);
+        $DB->insert_record('report_apocalypse', $activity->get_activity(), false);
```
you should be restructuring this to make use of the insert_records api calls instead of inserting a single record at a time.

tomdickman replied:
I removed this function and utilised the `insert_records` api call.

danmarsden commented:
The report_apocalypse_audits table isn't really needed - I probably would have just set a config var with the last time the audit ran so this value can be displayed on the main report view. If you really want to keep the table we probably need to implement a cleanup process to remove old records - otherwise this table is just going to fill up over time. (I'd probably just remove it.)

danmarsden commented on classes/audit_manager.php:
```diff
@@ -0,0 +1,376 @@
+        global $DB;
+
+        $record = new stdClass();
+        $date = new DateTime();
```
this seems weird - probably should just be:
$record->rundatetime = time();

tomdickman replied:
Mistake on my behalf, I didn't realise time() returned a Unix Epoch, I'll fix

tomdickman replied:
Fixed, removed superfluous code.

danmarsden commented on classes/output/renderer.php:
```diff
@@ -0,0 +1,111 @@
+            echo get_string('description', 'report_apocalypse');
+            echo $this->box_end();
+        }
+        $output = ob_get_contents();
```
you shouldn't be doing this unless you really have to - There might be a case for you to do this in the render_table process, but you should be able to just build $output and return it rather than using ob_get_contents()

tomdickman replied:
Fixed, removed the unnecessary buffering.

danmarsden commented on index.php:
```diff
@@ -32,108 +32,17 @@
-// Use table sort.
-$sort = $table->get_sql_sort();
-
-list($sql, $params) = report_apocalypse_sql($modules, $sort);
```
I would probably have preferred to see this file stay mostly as-is and just replace this call to get the sql with a call to get the sql from the new cached table instead of getting it directly from the db.

tomdickman replied:
When the table is rendered, the call to `get_audit_results_paginated` only pulls out a select amount of records from the stored table of most recent results. 
The sql and parameters for the audit are only built when `audit_manager::run_audit()` is called, which is invoked by the scheduled task.
Is this what you meant?

danmarsden replied:
No - I would have preferred you left view.php mostly as-is, created a separate process to populate a new table, and then just swap out the sql here with a sql call on your new table - make more sense? - I can live with some of the changes you've made, but you've over-complicated a lot of code that could have been left very simple.

I really don't see the point of the apocalypse_datetime class and a lot of your "helper" functions just aren't needed. - get_activityurl/get_activityname/get_htmlpresent/delete_records etc etc. You should just be calling API functions directly, not putting simple API calls into a helper function and calling your helper function.

tomdickman replied:
Ok I've gone back and removed the function/methods which aren't needed and pulled the API calls out.
The `renderer` is where all of the components are built for display, and the `audit_table` class just extends the `table_sql` class to build the html table from our new database `report_apocalypse` table, it doesn't run the audit.
The actual audit conduct is managed by the `audit_manager` and a call to it from the scheduled task, `scheduled_flash_audit`.
Is there anything else you want me to refactor?

tomdickman commented:
Thanks for the feedback @danmarsden ,

I have carried out the following changes in response:

- I have realigned the new table to the query data, while altering the `id` field to be `courseid` as it was proving troublesome trying to insert a column named id which may have duplicates, considering my primary key for indexing is also `id` and it seems Moodle parses out the `id` column.

- I have removed the necessity to get the course categories for each insert.

- I have removed the `apocalypse_datetime` class

- I have removed the `get_all_modules` function and replaced with inline call to API.

- I have kept `get_course_url_from_record` and `get_category_from_record` in, simply to avoid filling up my constructor with too much code for the `audit_activity` class.

- I have removed the unnecessary interface.

Let me know what you think of the changes,

Cheers

Tom

danmarsden commented:
hmm - maybe I'm reviewing the wrong code? - I can still see the apocalypse_dateteime class there... Also still a lot of abstraction/putting moodle api functions into functions when you should just be calling the api directly in-line.
